### PR TITLE
Add flag library

### PR DESCRIPTION
**Pull request by adleong**
One of the challenges with using the Cobra flag library is that it is difficult to differentiate between a flag that has been set explicitly to its default value and one that has been unset (and thus remains at its default value).  In particular, when processing CLI flags for the purposes of install and upgrade, this distinction is important because we want to persist the effects of flags which have been set.

Flag is an interface which describes a command line flag that affects the Helm Values used to render Helm charts.  This interface allows us to iterate over flags which have been set and apply their effects to the Values.  To see how this library is to be used, see https://github.com/linkerd/linkerd2/pull/5005

Signed-off-by: Alex Leong <alex@buoyant.io>

### CODE REVIEW

File: cli/flag/flag.go [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+	flagSet.DurationVar(&flag.Value, name, defaultValue, description)
+	return &flag
+}
+
+// Apply executes the stored apply function on the given Values.
+func (flag *UintFlag) Apply(values *charts.Values) error {
+	return flag.apply(values, flag.Value)
+}
+
+// IsSet returns true iff the Flag has been explicitly set with a value.
```
Comment (zaharidichev) [inline cli/flag/flag.go:159]:
```suggestion
// IsSet returns true if the Flag has been explicitly set with a value.
```

File: cli/flag/flag.go [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+func (flag *UintFlag) Name() string {
+	return flag.name
+}
+
+// Apply executes the stored apply function on the given Values.
+func (flag *Int64Flag) Apply(values *charts.Values) error {
+	return flag.apply(values, flag.Value)
+}
+
+// IsSet returns true iff the Flag has been explicitly set with a value.
```
Comment (zaharidichev) [inline cli/flag/flag.go:174]:
```suggestion
// IsSet returns true if the Flag has been explicitly set with a value.
```

File: cli/flag/flag.go [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+func (flag *Int64Flag) Name() string {
+	return flag.name
+}
+
+// Apply executes the stored apply function on the given Values.
+func (flag *StringFlag) Apply(values *charts.Values) error {
+	return flag.apply(values, flag.Value)
+}
+
+// IsSet returns true iff the Flag has been explicitly set with a value.
```
Comment (zaharidichev) [inline cli/flag/flag.go:189]:
```suggestion
// IsSet returns true if the Flag has been explicitly set with a value.
```

File: cli/flag/flag.go [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+func (flag *BoolFlag) Name() string {
+	return flag.name
+}
+
+// Apply executes the stored apply function on the given Values.
+func (flag *DurationFlag) Apply(values *charts.Values) error {
+	return flag.apply(values, flag.Value)
+}
+
+// IsSet returns true iff the Flag has been explicitly set with a value.
```
Comment (zaharidichev) [inline cli/flag/flag.go:234]:
```suggestion
// IsSet returns true if the Flag has been explicitly set with a value.
```

File: cli/flag/flag.go [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+func (flag *StringSliceFlag) Name() string {
+	return flag.name
+}
+
+// Apply executes the stored apply function on the given Values.
+func (flag *BoolFlag) Apply(values *charts.Values) error {
+	return flag.apply(values, flag.Value)
+}
+
+// IsSet returns true iff the Flag has been explicitly set with a value.
```
Comment (zaharidichev) [inline cli/flag/flag.go:219]:
```suggestion
// IsSet returns true if the Flag has been explicitly set with a value.
```

File: cli/flag/flag.go [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+func (flag *StringFlag) Name() string {
+	return flag.name
+}
+
+// Apply executes the stored apply function on the given Values.
+func (flag *StringSliceFlag) Apply(values *charts.Values) error {
+	return flag.apply(values, flag.Value)
+}
+
+// IsSet returns true iff the Flag has been explicitly set with a value.
```
Comment (zaharidichev) [inline cli/flag/flag.go:204]:
```suggestion
// IsSet returns true if the Flag has been explicitly set with a value.
```

File: cli/flag/flag.go [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+		flagSet: flagSet,
+		apply:   apply,
+	}
+	flagSet.UintVar(&flag.Value, name, defaultValue, description)
+	return &flag
+}
+
+// NewInt64Flag creates a new int64 typed Flag that executes the given function
+// when applied.  The flag is attached to the given FlagSet.
+func NewInt64Flag(flagSet *pflag.FlagSet, name string, defaultValue int64, description string, apply func(values *charts.Values, value int64) error) *Int64Flag {
```
Comment (zaharidichev) [inline cli/flag/flag.go:84]: NIT: Maybe you can consider using type aliases for the apply functions to make the signatures a bit more concise

**Review by zaharidichev — APPROVED** [review]
Nice abstraction!

**Review by kleimkuhler — APPROVED** [review]
Looks good!

**Pull request merged**